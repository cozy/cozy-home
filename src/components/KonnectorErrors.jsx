import PropTypes from 'prop-types'
import React from 'react'
import flow from 'lodash/flow'
import keyBy from 'lodash/keyBy'
import { connect } from 'react-redux'
import { useClient, models } from 'cozy-client'
import { useNavigate } from 'react-router-dom'

import AppIcon from 'cozy-ui/transpiled/react/AppIcon'
import Button from 'cozy-ui/transpiled/react/Button'
import CrossButton from 'cozy-ui/transpiled/react/Icons/Cross'
import Divider from 'cozy-ui/transpiled/react/MuiCozyTheme/Divider'
import Icon from 'cozy-ui/transpiled/react/Icon'
import IconButton from 'cozy-ui/transpiled/react/IconButton'
import Infos from 'cozy-ui/transpiled/react/Infos'
import InfosCarrousel from 'cozy-ui/transpiled/react/InfosCarrousel'
import Typography from 'cozy-ui/transpiled/react/Typography'
import useBreakpoints from 'cozy-ui/transpiled/react/hooks/useBreakpoints'
import { Media, Bd } from 'cozy-ui/transpiled/react/Media'
import { getErrorLocaleBound, KonnectorJobError } from 'cozy-harvest-lib'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import {
  getAccountsWithErrors,
  getInstalledKonnectors,
  getTriggersInError
} from 'reducers/index'
import ReactMarkdownWrapper from 'components/ReactMarkdownWrapper'
import homeConfig from 'config/home.json'

const {
  triggers: { triggers: triggersModel, triggerStates: triggerStatesModel },
  accounts: accountsModel
} = models

const muteTrigger = async (client, trigger, accountsById) => {
  const accountId = triggersModel.getAccountId(trigger)
  const initialAccount = accountsById[accountId]
  const errorType = triggerStatesModel.getLastErrorType(trigger)
  const account = accountsModel.muteError(initialAccount, errorType)

  await client.save(account)
}

const getKonnectorSlug = konnector => konnector.slug

// TODO, use directly Infos prop dismissAction when
// https://github.com/cozy/cozy-ui/pull/1724 is merged
// Here we need the aria-label for tests, and IconButton for style
const InfosDismissButton = ({ onClick }) => {
  const { t } = useI18n()
  return (
    <IconButton
      aria-label={t('connector.mute')}
      onClick={onClick}
      style={dismissButtonStyle}
    >
      <Icon icon={CrossButton} size={12} />
    </IconButton>
  )
}

const dismissButtonStyle = {
  position: 'absolute',
  top: '0.5rem',
  right: '0.5rem'
}

const KonnectorError = ({
  trigger,
  triggerErrors,
  index,
  konnectorsBySlug,
  accountsById
}) => {
  const client = useClient()
  const { t, lang } = useI18n()
  const { isMobile } = useBreakpoints()
  const errorType = triggerStatesModel.getLastErrorType(trigger)
  const konnError = new KonnectorJobError(errorType)
  const konnectorSlug = triggersModel.getKonnector(trigger)
  const konnectorAccount = triggersModel.getAccountId(trigger)
  const konnector = konnectorsBySlug[konnectorSlug]
  const navigate = useNavigate()

  const errorTitle = getErrorLocaleBound(konnError, konnector, lang, 'title')

  const errorDescription = getErrorLocaleBound(
    konnError,
    konnector,
    lang,
    'description'
  )

  const handleDismiss = () => {
    muteTrigger(client, trigger, accountsById)
  }

  return (
    <Infos
      theme="danger"
      key={trigger._id}
      description={
        <>
          <InfosDismissButton onClick={handleDismiss} />
          <Media>
            <AppIcon
              alt={t('app.logo.alt', { name: konnectorSlug })}
              app={konnectorSlug}
              type="konnector"
              className="u-w-2 u-h-2 u-w-1-half-s u-h-1-half-s u-mr-1"
            />
            <Bd>
              <Typography variant="caption" color="error">
                {konnector.name}
              </Typography>
              <Typography variant="h5" className="u-error">
                {triggerErrors.length > 1
                  ? `(${index + 1}/${triggerErrors.length}) `
                  : null}
                {errorTitle}
              </Typography>
            </Bd>
          </Media>
          <div>
            <Typography
              component="div"
              variant="body1"
              className="u-fz-small-m"
            >
              <ReactMarkdownWrapper source={errorDescription} />
            </Typography>
          </div>
        </>
      }
      action={
        <Button
          theme="secondary"
          label={t('fix_konnector_error')}
          className="u-mh-0"
          size={isMobile ? 'small' : 'normal'}
          onClick={() =>
            navigate(`/connected/${konnectorSlug}/accounts/${konnectorAccount}`)
          }
        />
      }
    />
  )
}

export const KonnectorErrors = ({
  triggersInError,
  accountsWithErrors,
  installedKonnectors
}) => {
  const accountsWithErrorsById = keyBy(accountsWithErrors, '_id')
  const installedKonnectorsBySlug = keyBy(installedKonnectors, getKonnectorSlug)
  const nonMutedTriggerErrors = triggersInError.filter(trigger => {
    const errorType = triggerStatesModel.getLastErrorType(trigger)
    const accountId = triggersModel.getAccountId(trigger)
    const account = accountsWithErrorsById[accountId]
    const konnectorSlug = triggersModel.getKonnector(trigger)
    const hasInstalledKonnector =
      installedKonnectors &&
      installedKonnectors.some(({ slug }) => slug === konnectorSlug)

    return (
      homeConfig.displayedErrorTypes.includes(errorType) &&
      hasInstalledKonnector &&
      account &&
      !triggersModel.isLatestErrorMuted(trigger, account)
    )
  })

  return nonMutedTriggerErrors.length > 0 ? (
    <div className="KonnectorErrors">
      <InfosCarrousel theme="danger">
        {nonMutedTriggerErrors.map((trigger, index) => (
          <KonnectorError
            key={trigger._id}
            trigger={trigger}
            triggerErrors={nonMutedTriggerErrors}
            konnectorsBySlug={installedKonnectorsBySlug}
            accountsById={accountsWithErrorsById}
            index={index}
          />
        ))}
      </InfosCarrousel>
    </div>
  ) : (
    <Divider className="u-mv-0" />
  )
}

KonnectorErrors.propTypes = {
  accountsWithErrors: PropTypes.array.isRequired,
  installedKonnectors: PropTypes.arrayOf(PropTypes.object.isRequired),
  triggersInError: PropTypes.array.isRequired
}

const mapStateToProps = state => {
  return {
    accountsWithErrors: getAccountsWithErrors(state),
    installedKonnectors: getInstalledKonnectors(state),
    triggersInError: getTriggersInError(state)
  }
}

export default flow(connect(mapStateToProps))(KonnectorErrors)
