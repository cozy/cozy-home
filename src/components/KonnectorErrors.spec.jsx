import React from 'react'
import MockDate from 'mockdate'
import { KonnectorErrors } from './KonnectorErrors'
import AppLike from '../../test/AppLike'
import { render, fireEvent } from '@testing-library/react'

jest.mock('cozy-ui/transpiled/react/AppIcon', () => () => null)

describe('KonnectorErrors', () => {
  const MOCKED_DATE = '2020-01-08T09:49:23.589Z'
  beforeAll(() => {
    MockDate.set(MOCKED_DATE)
  })

  afterAll(() => {
    jest.restoreAllMocks()
    MockDate.reset()
  })

  const mockClient = {
    save: jest.fn()
  }

  const mockHistory = {}

  const DEFAULT_INSTALLED_KONNECTORS = [
    { slug: 'test', name: 'Test Konnector' }
  ]

  const setup = ({
    triggersInError = [],
    accountsWithErrors = [],
    installedKonnectors = DEFAULT_INSTALLED_KONNECTORS
  } = {}) => {
    const root = render(
      <AppLike client={mockClient}>
        <KonnectorErrors
          triggersInError={triggersInError}
          accountsWithErrors={accountsWithErrors}
          installedKonnectors={installedKonnectors}
          history={mockHistory}
        />
      </AppLike>
    )
    return { root }
  }

  it('should render divider when there are no errors', () => {
    const { root } = setup()
    expect(root.container).toMatchSnapshot()
  })

  it('should render divider when there are errors but no installed konnector', () => {
    const triggersInError = [
      {
        _id: '2',
        worker: 'konnector',
        current_state: {
          last_error: 'MUTED_ERROR',
          last_success: '2019-10-01T00:48:01.404911778Z'
        },
        message: {
          konnector: 'test',
          account: '456'
        }
      }
    ]
    const { root } = setup({ triggersInError })

    expect(root.container).toMatchSnapshot()
  })

  it('should render divider when all errors are muted', () => {
    const triggersInError = [
      {
        _id: '2',
        worker: 'konnector',
        current_state: {
          last_error: 'MUTED_ERROR',
          last_success: '2019-10-01T00:48:01.404911778Z'
        },
        message: {
          konnector: 'test',
          account: '456'
        }
      }
    ]
    const accountsWithErrors = [
      {
        _id: '456',
        mutedErrors: [
          {
            type: 'MUTED_ERROR',
            mutedAt: '2019-12-01T00:48:01.404911778Z'
          }
        ]
      }
    ]
    const installedKonnectors = [{ slug: 'test', name: 'Test Konnector' }]

    const { root } = setup({
      triggersInError,
      accountsWithErrors,
      installedKonnectors
    })

    expect(root.container).toMatchSnapshot()
  })

  it('should render active errors', async () => {
    const triggersInError = [
      {
        _id: '1',
        worker: 'konnector',
        current_state: {
          last_error: 'LOGIN_FAILED'
        },
        message: {
          konnector: 'test',
          account: '123'
        }
      },
      {
        _id: '2',
        worker: 'konnector',
        current_state: {
          last_error: 'LOGIN_FAILED.NEEDS_SECRET', // this one is muted
          last_success: '2019-10-01T00:48:01.404911778Z'
        },
        message: {
          konnector: 'test',
          account: '456'
        }
      },
      {
        _id: '3',
        worker: 'konnector',
        current_state: {
          last_error: 'USER_ACTION_NEEDED'
        },
        message: {
          konnector: 'test',
          account: '123'
        }
      },
      {
        _id: '4',
        worker: 'konnector',
        current_state: {
          last_error: 'VENDOR_DOWN' // This type of error is not displayed
        },
        message: {
          konnector: 'test',
          account: '123'
        }
      }
    ]
    const accountsWithErrors = [
      { _id: '123', mutedErrors: [] },
      {
        _id: '456',
        mutedErrors: [
          {
            type: 'LOGIN_FAILED.NEEDS_SECRET',
            mutedAt: '2019-12-01T00:48:01.404911778Z'
          }
        ]
      }
    ]
    const installedKonnectors = [{ slug: 'test', name: 'Test Konnector' }]
    const { root } = setup({
      triggersInError,
      accountsWithErrors,
      installedKonnectors
    })

    expect(
      root.getByText('(1/2) Incorrect or expired credentials')
    ).toBeTruthy()

    const dismissButton = root.getByLabelText('Mute error')
    await fireEvent.click(dismissButton)

    expect(mockClient.save).toHaveBeenCalledWith({
      _id: '123',
      mutedErrors: [{ mutedAt: MOCKED_DATE, type: 'LOGIN_FAILED' }]
    })
  })

  it('should hide errors when the konnector or account is missing', () => {
    const triggersInError = [
      {
        _id: '1',
        worker: 'konnector',
        current_state: {
          last_error: 'LOGIN_FAILED'
        },
        message: {
          konnector: 'uninstalled',
          account: '123'
        }
      },
      {
        _id: '2',
        worker: 'konnector',
        current_state: {
          last_error: 'LOGIN_FAILED'
        },
        message: {
          konnector: 'uninstalled',
          account: 'no-account'
        }
      }
    ]
    const accountsWithErrors = [{ _id: '123', mutedErrors: [] }]
    const installedKonnectors = []
    const { root } = setup({
      triggersInError,
      accountsWithErrors,
      installedKonnectors
    })
    expect(root.container).toMatchSnapshot()
  })

  it('should not show slide indicator with only one slide', () => {
    const triggersInError = [
      {
        _id: '1',
        worker: 'konnector',
        current_state: {
          last_error: 'LOGIN_FAILED'
        },
        message: {
          konnector: 'test',
          account: '123'
        }
      }
    ]
    const accountsWithErrors = [
      { _id: '123', mutedErrors: [] },
      {
        _id: '456',
        mutedErrors: [
          {
            type: 'LOGIN_FAILED.NEEDS_SECRET',
            mutedAt: '2019-12-01T00:48:01.404911778Z'
          }
        ]
      }
    ]
    const installedKonnectors = [{ slug: 'test', name: 'Test Konnector' }]
    const { root } = setup({
      triggersInError,
      accountsWithErrors,
      installedKonnectors
    })

    // 1/1 is not displayed
    expect(root.getByText('Incorrect or expired credentials'))
  })
})
