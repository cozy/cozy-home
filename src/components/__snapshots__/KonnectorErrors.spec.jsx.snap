// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KonnectorErrors should hide errors when the konnector or account is missing 1`] = `
<div>
  <hr
    class="MuiDivider-root u-mv-0"
  />
</div>
`;

exports[`KonnectorErrors should render divider when all errors are muted 1`] = `
<div>
  <hr
    class="MuiDivider-root u-mv-0"
  />
</div>
`;

exports[`KonnectorErrors should render divider when there are errors but no installed konnector 1`] = `
<div>
  <hr
    class="MuiDivider-root u-mv-0"
  />
</div>
`;

exports[`KonnectorErrors should render divider when there are no errors 1`] = `
<div>
  <hr
    class="MuiDivider-root u-mv-0"
  />
</div>
`;
